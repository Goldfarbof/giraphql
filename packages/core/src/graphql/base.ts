import { GraphQLType } from 'graphql';
import BasePlugin from '../plugin';
import BuildCache from '../build-cache';

export default abstract class BaseType<
  Types extends GiraphQLSchemaTypes.TypeInfo,
  Name extends string | keyof Types['Input'] | keyof Types['Output'],
  Shape,
  InputShape = Shape,
  MatchShape = Shape
> {
  typename: Name;

  shape?: Shape;

  inputShape?: InputShape;

  matchShape?: MatchShape;

  abstract kind: 'Object' | 'Union' | 'Interface' | 'Enum' | 'Scalar' | 'InputObject';

  constructor(name: Name, shape?: Shape) {
    this.typename = name;
    this.shape = shape;
  }

  abstract buildType(cache: BuildCache<Types>, plugins: BasePlugin<Types>[]): GraphQLType;
}
