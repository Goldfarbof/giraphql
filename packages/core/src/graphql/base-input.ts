import BaseType from './base';

export default abstract class BaseInputType<Shape = unknown, InputShape = Shape> extends BaseType<
  Shape
> {
  shape?: Shape;

  inputShape?: InputShape;

  abstract kind: 'Enum' | 'Scalar' | 'InputObject';
}
