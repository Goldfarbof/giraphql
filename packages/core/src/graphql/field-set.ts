import { FieldsShape } from '../types';

export default class FieldSet<Types extends GiraphQLSchemaTypes.TypeInfo, Shape = unknown> {
  kind: 'FieldSet' = 'FieldSet';

  forType: string;

  shape: FieldsShape<any, any>;

  constructor(type: string, shape: FieldsShape<Types, Shape>) {
    this.forType = type;
    this.shape = shape;
  }
}
