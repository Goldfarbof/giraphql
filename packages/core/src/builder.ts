/* eslint-disable no-dupe-class-members */
import { GraphQLSchema, GraphQLScalarType, GraphQLDirective } from 'graphql';
import {
  ImplementedType,
  EnumValues,
  InputFields,
  InputShapeFromFields,
  ObjectName,
  InterfaceName,
  ScalarName,
  ResolverMap,
  CompatibleInterfaceParam,
  FieldMap,
  ObjectFieldsShape,
  QueryFieldsShape,
  MutationFieldsShape,
  SubscriptionFieldsShape,
  InterfaceFieldsShape,
  ObjectFieldThunk,
  InterfaceFieldThunk,
  QueryFieldThunk,
  MutationFieldThunk,
  SubscriptionFieldThunk,
} from './types';
import ObjectType from './graphql/object';
import UnionType from './graphql/union';
import InputObjectType from './graphql/input';
import InterfaceType from './graphql/interface';
import EnumType from './graphql/enum';
import ScalarType from './graphql/scalar';
import BuildCache from './build-cache';
import {
  InputFieldBuilder,
  ObjectFieldBuilder,
  QueryFieldBuilder,
  MutationFieldBuilder,
  SubscriptionFieldBuilder,
  InterfaceFieldBuilder,
} from '.';
import { BasePlugin, mergePlugins } from './plugins';
import QueryType from './graphql/query';
import MutationType from './graphql/mutation';
import SubscriptionType from './graphql/subscription';

export default class SchemaBuilder<Types extends GiraphQLSchemaTypes.TypeInfo> {
  private plugin: Required<BasePlugin>;

  private types = new Map<string, ImplementedType>();

  private pendingFields = new Map<string, FieldMap[]>();

  private fields = new Map<string, FieldMap[]>();

  constructor(options: { plugins?: BasePlugin[] } = {}) {
    this.plugin = mergePlugins(options.plugins ?? []);
  }

  objectType<
    Interfaces extends CompatibleInterfaceParam<Types, Types['Object'][Type]>[],
    Type extends ObjectName<Types>
  >(
    name: Type,
    options:
      | GiraphQLSchemaTypes.ObjectTypeOptions<Types, Types['Object'][Type]>
      | GiraphQLSchemaTypes.ObjectTypeWithInterfaceOptions<
          Types,
          Types['Object'][Type],
          Interfaces
        >,
    shape?: ObjectFieldsShape<Types, Types['Object'][Type]>,
  ) {
    const type = this.addType(
      new ObjectType<Types>(
        name,
        options as GiraphQLSchemaTypes.ObjectTypeOptions<Types, Types['Object'][Type]>,
      ),
    );

    if (shape) {
      this.addFields(name, shape(new ObjectFieldBuilder(name)));
    }

    if (options.shape) {
      this.addFields(name, options.shape(new ObjectFieldBuilder(name)));
    }

    return type;
  }

  objectFields<Type extends ObjectName<Types>>(
    name: Type,
    shape: ObjectFieldsShape<Types, Types['Object'][Type]>,
  ) {
    return this.addFields(name, shape(new ObjectFieldBuilder(name)));
  }

  objectField<Type extends ObjectName<Types>>(
    name: Type,
    fieldName: string,
    field: ObjectFieldThunk<Types, Types['Object'][Type]>,
  ) {
    return this.addFields(name, { [fieldName]: field(new ObjectFieldBuilder(name)) });
  }

  queryType(options: GiraphQLSchemaTypes.QueryTypeOptions<Types>, shape?: QueryFieldsShape<Types>) {
    const type = this.addType(new QueryType<Types>(options));

    if (shape) {
      this.addFields('Query', shape(new QueryFieldBuilder()));
    }

    if (options.shape) {
      this.addFields('Query', options.shape(new QueryFieldBuilder()));
    }

    return type;
  }

  queryFields(shape: QueryFieldsShape<Types>) {
    return this.addFields('Query', shape(new QueryFieldBuilder()));
  }

  queryField(name: string, field: QueryFieldThunk<Types>) {
    return this.addFields('Query', { [name]: field(new QueryFieldBuilder()) });
  }

  mutationType(
    options: GiraphQLSchemaTypes.MutationTypeOptions<Types>,
    shape?: MutationFieldsShape<Types>,
  ) {
    const type = this.addType(new MutationType<Types>(options));

    if (shape) {
      this.addFields('Mutation', shape(new MutationFieldBuilder()));
    }

    if (options.shape) {
      this.addFields('Mutation', options.shape(new MutationFieldBuilder()));
    }

    return type;
  }

  mutationFields(shape: MutationFieldsShape<Types>) {
    return this.addFields('Mutation', shape(new MutationFieldBuilder()));
  }

  mutationField(name: string, field: MutationFieldThunk<Types>) {
    return this.addFields('Mutation', { [name]: field(new MutationFieldBuilder()) });
  }

  subscriptionType(
    options: GiraphQLSchemaTypes.SubscriptionTypeOptions<Types>,
    shape?: SubscriptionFieldsShape<Types>,
  ) {
    const type = this.addType(new SubscriptionType<Types>(options));

    if (shape) {
      this.addFields('Subscription', shape(new SubscriptionFieldBuilder()));
    }

    if (options.shape) {
      this.addFields('Subscription', options.shape(new SubscriptionFieldBuilder()));
    }

    return type;
  }

  subscriptionFields(shape: SubscriptionFieldsShape<Types>) {
    return this.addFields('Subscription', shape(new SubscriptionFieldBuilder()));
  }

  subscriptionField(name: string, field: SubscriptionFieldThunk<Types>) {
    return this.addFields('Subscription', { [name]: field(new SubscriptionFieldBuilder()) });
  }

  args<Shape extends InputFields<Types>>(
    shape: (t: GiraphQLSchemaTypes.InputFieldBuilder<Types>) => Shape,
  ) {
    return shape(new InputFieldBuilder<Types>());
  }

  interfaceType<Type extends InterfaceName<Types>>(
    name: Type,
    options: GiraphQLSchemaTypes.InterfaceTypeOptions<Types, Types['Interface'][Type]>,
    shape?: InterfaceFieldsShape<Types, Types['Interface'][Type]>,
  ) {
    const type = this.addType(new InterfaceType<Types, Type>(name, options));

    if (shape) {
      this.addFields(name, shape(new InterfaceFieldBuilder(name)));
    }

    if (options.shape) {
      this.addFields(name, options.shape(new InterfaceFieldBuilder(name)));
    }

    return type;
  }

  interfaceFields<Type extends InterfaceName<Types>>(
    name: Type,
    shape: InterfaceFieldsShape<Types, Types['Interface'][Type]>,
  ) {
    return this.addFields(name, shape(new InterfaceFieldBuilder(name)));
  }

  interfaceField<Type extends InterfaceName<Types>>(
    name: Type,
    fieldName: string,
    field: InterfaceFieldThunk<Types, Types['Interface'][Type]>,
  ) {
    return this.addFields(name, { [fieldName]: field(new InterfaceFieldBuilder(name)) });
  }

  unionType<Member extends ObjectName<Types>, Name extends string>(
    name: Name,
    options: GiraphQLSchemaTypes.UnionOptions<Types, Member>,
  ) {
    return this.addType(new UnionType<Types, Member>(name, options));
  }

  enumType<Name extends string, Values extends EnumValues>(
    name: Name,
    options: GiraphQLSchemaTypes.EnumTypeOptions<Values>,
  ) {
    return this.addType(new EnumType(name, options));
  }

  scalarType<Name extends ScalarName<Types>>(
    name: Name,
    options: GiraphQLSchemaTypes.ScalarOptions<
      Types['Scalar'][Name]['Input'],
      Types['Scalar'][Name]['Output']
    >,
  ) {
    return this.addType(
      new ScalarType<Types['Scalar'][Name]['Input'], Types['Scalar'][Name]['Output']>(
        name,
        options,
      ),
    );
  }

  addScalarType<Name extends ScalarName<Types>>(name: Name, scalar: GraphQLScalarType) {
    return this.addType(
      new ScalarType<Types['Scalar'][Name]['Input'], Types['Scalar'][Name]['Output']>(name, {
        description: scalar.description ?? undefined,
        serialize: scalar.serialize,
        parseLiteral: scalar.parseLiteral,
        parseValue: scalar.parseValue,
        extensions: scalar.extensions ?? undefined,
      }),
    );
  }

  inputType<Name extends string, Fields extends InputFields<Types>>(
    name: Name,
    options: GiraphQLSchemaTypes.InputTypeOptions<Types, Fields>,
  ): InputObjectType<Types, InputShapeFromFields<Types, Fields, undefined>, Name>;

  inputType<Name extends string, Fields extends InputFields<Types>>(
    name: Name,
    options: Omit<GiraphQLSchemaTypes.InputTypeOptions<Types, Fields>, 'shape'>,
    shape: (t: GiraphQLSchemaTypes.InputFieldBuilder<Types>) => Fields,
  ): InputObjectType<Types, InputShapeFromFields<Types, Fields, undefined>, Name>;

  inputType<Name extends string, Fields extends InputFields<Types>>(
    name: Name,
    options: GiraphQLSchemaTypes.InputTypeOptions<Types, Fields>,
    shape?: (t: GiraphQLSchemaTypes.InputFieldBuilder<Types>) => Fields,
  ): InputObjectType<Types, InputShapeFromFields<Types, Fields, undefined>, Name> {
    if (shape && options.shape) {
      throw new Error(`Received multiple field shape functions for InputObjectType ${name}`);
    }

    return this.addType(
      new InputObjectType<Types, InputShapeFromFields<Types, Fields, undefined>, Name>(name, {
        ...options,
        shape: shape ?? options.shape,
      }),
    );
  }

  // Temporarily alias old names for compatibility
  createQueryType = this.queryType;

  createMutationType = this.mutationType;

  createSubscriptionType = this.subscriptionType;

  createObjectType = this.objectType;

  createInterfaceType = this.interfaceType;

  createQueryFields = this.queryFields;

  createMutationFields = this.mutationFields;

  createSubscriptionFields = this.subscriptionFields;

  createObjectFields = this.objectFields;

  createInterfaceFields = this.interfaceFields;

  createUnionType = this.unionType;

  createEnumType = this.enumType;

  createScalarType = this.addScalarType;

  createInputType = this.inputType;

  createArgs = this.args;

  toSchema({
    directives,
    extensions,
    mocks,
  }: {
    directives?: readonly GraphQLDirective[];
    extensions?: Record<string, unknown>;
    mocks?: ResolverMap;
  } = {}) {
    this.plugin.beforeBuild(this);

    if (this.pendingFields.size) {
      throw new Error(
        `Fields defined without defining type (${[...this.pendingFields.keys()].join(', ')}).`,
      );
    }

    const buildCache = new BuildCache([...this.types.values()], this.plugin, {
      fieldDefinitions: this.fields,
      mocks,
    });

    buildCache.buildAll();

    const builtTypes = [...buildCache.types.values()].map((entry) => entry.built);

    const schema = new GraphQLSchema({
      query: buildCache.has('Query')
        ? buildCache.getEntryOfType('Query', 'Query').built
        : undefined,
      mutation: buildCache.has('Mutation')
        ? buildCache.getEntryOfType('Mutation', 'Mutation').built
        : undefined,
      subscription: buildCache.has('Subscription')
        ? buildCache.getEntryOfType('Subscription', 'Subscription').built
        : undefined,
      extensions,
      directives: directives as GraphQLDirective[],
      types: builtTypes,
    });

    this.plugin.afterBuild(schema, this);

    return schema;
  }

  addType<T extends ImplementedType>(type: T) {
    this.types.set(type.typename, type);

    this.plugin.onType(type, this);

    const pending = this.pendingFields.get(type.typename);

    if (pending) {
      this.pendingFields.delete(type.typename);
      for (const fields of pending) {
        this.addFields(type.typename, fields);
      }
    }

    return type;
  }

  addFields(typename: string, fields: FieldMap) {
    const type = this.types.get(typename);

    if (type) {
      if (this.fields.has(typename)) {
        this.fields.get(typename)!.push(fields);
      } else {
        this.fields.set(typename, [fields]);
      }

      Object.keys(fields).forEach((fieldName) => {
        this.plugin.onField(type, fieldName, fields[fieldName], this);
      });
    } else if (this.pendingFields.has(typename)) {
      this.pendingFields.get(typename)!.push(fields);
    } else {
      this.pendingFields.set(typename, [fields]);
    }

    return fields;
  }
}
