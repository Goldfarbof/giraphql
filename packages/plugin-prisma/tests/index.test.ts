import { gql } from 'apollo-server';
import { execute, printSchema } from 'graphql';
import { prisma } from './example/builder';
import schema from './example/schema';

let queries: unknown[] = [];
prisma.$use((params, next) => {
  queries.push(params);

  return next(params);
});

describe('prisma', () => {
  afterEach(() => {
    queries = [];
  });

  afterAll(async () => {
    await prisma.$disconnect();
  });

  it('generates schema', () => {
    expect(printSchema(schema)).toMatchSnapshot();
  });

  it('queries for single item', async () => {
    const query = gql`
      query {
        me {
          id
        }
      }
    `;

    const result = await execute({
      schema,
      document: query,
      contextValue: { user: { id: 1 } },
    });

    expect(result).toMatchInlineSnapshot(`
      Object {
        "data": Object {
          "me": Object {
            "id": "VXNlcjox",
          },
        },
      }
      `);

    expect(queries).toMatchInlineSnapshot(`
      Array [
        Object {
          "action": "findUnique",
          "args": Object {
            "where": Object {
              "id": 1,
            },
          },
          "dataPath": Array [],
          "model": "User",
          "runInTransaction": false,
        },
      ]
    `);
  });

  it('queries for a list of records', async () => {
    const query = gql`
      query {
        users {
          id
        }
      }
    `;

    const result = await execute({
      schema,
      document: query,
      contextValue: { user: { id: 1 } },
    });

    expect(result).toMatchInlineSnapshot(`
      Object {
        "data": Object {
          "users": Array [
            Object {
              "id": "VXNlcjox",
            },
            Object {
              "id": "VXNlcjoy",
            },
          ],
        },
      }
    `);

    expect(queries).toMatchInlineSnapshot(`
Array [
  Object {
    "action": "findMany",
    "args": Object {
      "take": 2,
    },
    "dataPath": Array [],
    "model": "User",
    "runInTransaction": false,
  },
]
`);
  });

  it('queries for record with nested relations', async () => {
    const query = gql`
      query {
        me {
          posts {
            author {
              profile {
                bio
              }
            }
          }
        }
      }
    `;

    const result = await execute({
      schema,
      document: query,
      contextValue: { user: { id: 1 } },
    });

    expect(result).toMatchSnapshot();

    expect(queries).toMatchInlineSnapshot(`
Array [
  Object {
    "action": "findUnique",
    "args": Object {
      "include": Object {
        "posts": Object {
          "include": Object {
            "author": Object {
              "include": Object {
                "profile": true,
              },
            },
            "comments": Object {
              "include": Object {
                "author": true,
              },
            },
          },
          "orderBy": Object {
            "createdAt": "desc",
          },
          "take": 10,
        },
      },
      "where": Object {
        "id": 1,
      },
    },
    "dataPath": Array [],
    "model": "User",
    "runInTransaction": false,
  },
]
`);
  });

  it('queries for list with nested relations', async () => {
    const query = gql`
      query {
        users {
          name
          profile {
            bio
          }
          posts {
            id
            author {
              profile {
                id
              }
            }
          }
        }
      }
    `;

    const result = await execute({
      schema,
      document: query,
      contextValue: { user: { id: 1 } },
    });

    expect(result).toMatchSnapshot();

    expect(queries).toMatchInlineSnapshot(`
Array [
  Object {
    "action": "findMany",
    "args": Object {
      "include": Object {
        "posts": Object {
          "include": Object {
            "author": Object {
              "include": Object {
                "profile": true,
              },
            },
            "comments": Object {
              "include": Object {
                "author": true,
              },
            },
          },
          "orderBy": Object {
            "createdAt": "desc",
          },
          "take": 10,
        },
        "profile": true,
      },
      "take": 2,
    },
    "dataPath": Array [],
    "model": "User",
    "runInTransaction": false,
  },
]
`);
  });
  it('queries with arguments and aliases', async () => {
    const query = gql`
      query {
        me {
          posts {
            id
            author {
              id
            }
          }
          oldestPosts: posts(oldestFirst: true) {
            id
            author {
              id
            }
          }
        }
      }
    `;

    const result = await execute({
      schema,
      document: query,
      contextValue: { user: { id: 1 } },
    });

    expect(result).toMatchSnapshot();

    expect(queries).toMatchInlineSnapshot(`
Array [
  Object {
    "action": "findUnique",
    "args": Object {
      "include": Object {
        "posts": Object {
          "include": Object {
            "author": true,
            "comments": Object {
              "include": Object {
                "author": true,
              },
            },
          },
          "orderBy": Object {
            "createdAt": "desc",
          },
          "take": 10,
        },
      },
      "where": Object {
        "id": 1,
      },
    },
    "dataPath": Array [],
    "model": "User",
    "runInTransaction": false,
  },
  Object {
    "action": "findMany",
    "args": Object {
      "include": Object {
        "author": true,
        "comments": Object {
          "include": Object {
            "author": true,
          },
        },
      },
      "orderBy": Object {
        "createdAt": "asc",
      },
      "take": 10,
      "where": Object {
        "authorId": 1,
      },
    },
    "dataPath": Array [],
    "model": "Post",
    "runInTransaction": false,
  },
]
`);
  });

  it('queries with variables and alieases', async () => {
    const query = gql`
      query ($oldest: Boolean) {
        me {
          posts {
            id
            author {
              id
            }
          }
          postIds: posts {
            id
          }
          oldestPosts: posts(oldestFirst: $oldest) {
            id
            author {
              id
            }
          }
        }
      }
    `;

    const result = await execute({
      schema,
      document: query,
      contextValue: { user: { id: 1 } },
      variableValues: {
        oldest: true,
      },
    });

    expect(result).toMatchSnapshot();

    expect(queries).toMatchInlineSnapshot(`
Array [
  Object {
    "action": "findUnique",
    "args": Object {
      "include": Object {
        "posts": Object {
          "include": Object {
            "author": true,
            "comments": Object {
              "include": Object {
                "author": true,
              },
            },
          },
          "orderBy": Object {
            "createdAt": "desc",
          },
          "take": 10,
        },
      },
      "where": Object {
        "id": 1,
      },
    },
    "dataPath": Array [],
    "model": "User",
    "runInTransaction": false,
  },
  Object {
    "action": "findMany",
    "args": Object {
      "include": Object {
        "author": true,
        "comments": Object {
          "include": Object {
            "author": true,
          },
        },
      },
      "orderBy": Object {
        "createdAt": "asc",
      },
      "take": 10,
      "where": Object {
        "authorId": 1,
      },
    },
    "dataPath": Array [],
    "model": "Post",
    "runInTransaction": false,
  },
]
`);
  });

  it('queries through non-prisma fields', async () => {
    const query = gql`
      query ($oldest: Boolean) {
        me {
          profileThroughManualLookup {
            user {
              profile {
                bio
              }
            }
          }
        }
      }
    `;

    const result = await execute({
      schema,
      document: query,
      contextValue: { user: { id: 1 } },
      variableValues: {
        oldest: true,
      },
    });

    expect(result).toMatchInlineSnapshot(`
Object {
  "data": Object {
    "me": Object {
      "profileThroughManualLookup": Object {
        "user": Object {
          "profile": Object {
            "bio": "Sequi minus inventore itaque similique et.",
          },
        },
      },
    },
  },
}
`);

    expect(queries).toMatchInlineSnapshot(`
      Array [
        Object {
          "action": "findUnique",
          "args": Object {
            "where": Object {
              "id": 1,
            },
          },
          "dataPath": Array [],
          "model": "User",
          "runInTransaction": false,
        },
        Object {
          "action": "findUnique",
          "args": Object {
            "select": Object {
              "profile": true,
            },
            "where": Object {
              "id": 1,
            },
          },
          "dataPath": Array [
            "select",
            "profile",
          ],
          "model": "User",
          "runInTransaction": false,
        },
        Object {
          "action": "findUnique",
          "args": Object {
            "include": Object {
              "profile": true,
            },
            "where": Object {
              "id": 1,
            },
          },
          "dataPath": Array [],
          "model": "User",
          "runInTransaction": false,
        },
      ]
    `);
  });

  it('queries with errors plugin', async () => {
    const query = gql`
      query {
        usersWithError {
          __typename
          ... on Error {
            message
          }
          ... on QueryUsersWithErrorSuccess {
            data {
              name
              profileWithErrors {
                ... on Error {
                  message
                }
                ... on UserProfileWithErrorsSuccess {
                  data {
                    bio
                    user {
                      id
                    }
                  }
                }
              }
            }
          }
        }
      }
    `;

    const result = await execute({
      schema,
      document: query,
      contextValue: { user: { id: 1 } },
      variableValues: {
        oldest: true,
      },
    });

    expect(result).toMatchInlineSnapshot(`
Object {
  "data": Object {
    "usersWithError": Object {
      "__typename": "QueryUsersWithErrorSuccess",
      "data": Array [
        Object {
          "name": "Maurine Ratke",
          "profileWithErrors": Object {
            "data": Object {
              "bio": "Sequi minus inventore itaque similique et.",
              "user": Object {
                "id": "VXNlcjox",
              },
            },
          },
        },
        Object {
          "name": "Nichole Koss",
          "profileWithErrors": null,
        },
      ],
    },
  },
}
`);

    expect(queries).toMatchInlineSnapshot(`
Array [
  Object {
    "action": "findMany",
    "args": Object {
      "include": Object {
        "profile": Object {
          "include": Object {
            "user": true,
          },
        },
      },
      "take": 2,
    },
    "dataPath": Array [],
    "model": "User",
    "runInTransaction": false,
  },
]
`);
  });
});
