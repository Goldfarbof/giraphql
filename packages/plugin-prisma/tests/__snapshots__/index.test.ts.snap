// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prisma generates schema 1`] = `
"interface Named {
  name: String
}

interface Node {
  id: ID!
}

type PageInfo {
  endCursor: String
  hasNextPage: Boolean!
  hasPreviousPage: Boolean!
  startCursor: String
}

type Post {
  author: User!
  content: String
  createdAt: String!
  id: ID!
  title: String!
}

type Profile {
  bio: String
  id: ID!
  user: User
}

type Query {
  me: User
  named: [Named]
  node(id: ID!): Node
  nodes(ids: [ID!]!): [Node]!
  userConnection(after: ID, before: ID, first: Int, last: Int): QueryUserConnection!
  userOrProfile: [UserOrProfile]
  users: [User!]!
}

type QueryUserConnection {
  edges: [QueryUserConnectionEdge]!
  pageInfo: PageInfo!
}

type QueryUserConnectionEdge {
  cursor: String!
  node: User!
}

type User implements Named & Node {
  email: String!
  id: ID!
  name: String
  posts(oldestFirst: Boolean): [Post!]!
  postsConnection(after: ID, before: ID, first: Int, last: Int, oldestFirst: Boolean): UserPostsConnection!
  profile: Profile!
  profileThroughManualLookup: Profile
}

union UserOrProfile = Profile | User

type UserPostsConnection {
  edges: [UserPostsConnectionEdge]!
  pageInfo: PageInfo!
}

type UserPostsConnectionEdge {
  cursor: String!
  node: Post!
}
"
`;

exports[`prisma queries for list with nested relations 1`] = `
Object {
  "data": Object {
    "users": Array [
      Object {
        "name": "Maurine Ratke",
        "posts": Array [
          Object {
            "author": Object {
              "profile": Object {
                "id": "1",
              },
            },
            "id": "250",
          },
          Object {
            "author": Object {
              "profile": Object {
                "id": "1",
              },
            },
            "id": "249",
          },
          Object {
            "author": Object {
              "profile": Object {
                "id": "1",
              },
            },
            "id": "248",
          },
          Object {
            "author": Object {
              "profile": Object {
                "id": "1",
              },
            },
            "id": "247",
          },
          Object {
            "author": Object {
              "profile": Object {
                "id": "1",
              },
            },
            "id": "246",
          },
          Object {
            "author": Object {
              "profile": Object {
                "id": "1",
              },
            },
            "id": "245",
          },
          Object {
            "author": Object {
              "profile": Object {
                "id": "1",
              },
            },
            "id": "244",
          },
          Object {
            "author": Object {
              "profile": Object {
                "id": "1",
              },
            },
            "id": "243",
          },
          Object {
            "author": Object {
              "profile": Object {
                "id": "1",
              },
            },
            "id": "242",
          },
          Object {
            "author": Object {
              "profile": Object {
                "id": "1",
              },
            },
            "id": "241",
          },
        ],
        "profile": Object {
          "bio": "Sequi minus inventore itaque similique et.",
        },
      },
      Object {
        "name": "Nichole Koss",
        "posts": Array [
          Object {
            "author": Object {
              "profile": Object {
                "id": "2",
              },
            },
            "id": "500",
          },
          Object {
            "author": Object {
              "profile": Object {
                "id": "2",
              },
            },
            "id": "499",
          },
          Object {
            "author": Object {
              "profile": Object {
                "id": "2",
              },
            },
            "id": "498",
          },
          Object {
            "author": Object {
              "profile": Object {
                "id": "2",
              },
            },
            "id": "497",
          },
          Object {
            "author": Object {
              "profile": Object {
                "id": "2",
              },
            },
            "id": "496",
          },
          Object {
            "author": Object {
              "profile": Object {
                "id": "2",
              },
            },
            "id": "495",
          },
          Object {
            "author": Object {
              "profile": Object {
                "id": "2",
              },
            },
            "id": "494",
          },
          Object {
            "author": Object {
              "profile": Object {
                "id": "2",
              },
            },
            "id": "493",
          },
          Object {
            "author": Object {
              "profile": Object {
                "id": "2",
              },
            },
            "id": "492",
          },
          Object {
            "author": Object {
              "profile": Object {
                "id": "2",
              },
            },
            "id": "491",
          },
        ],
        "profile": Object {
          "bio": "Ut quo accusantium fuga veritatis.",
        },
      },
    ],
  },
}
`;

exports[`prisma queries for record with nested relations 1`] = `
Object {
  "data": Object {
    "me": Object {
      "posts": Array [
        Object {
          "author": Object {
            "profile": Object {
              "bio": "Sequi minus inventore itaque similique et.",
            },
          },
        },
        Object {
          "author": Object {
            "profile": Object {
              "bio": "Sequi minus inventore itaque similique et.",
            },
          },
        },
        Object {
          "author": Object {
            "profile": Object {
              "bio": "Sequi minus inventore itaque similique et.",
            },
          },
        },
        Object {
          "author": Object {
            "profile": Object {
              "bio": "Sequi minus inventore itaque similique et.",
            },
          },
        },
        Object {
          "author": Object {
            "profile": Object {
              "bio": "Sequi minus inventore itaque similique et.",
            },
          },
        },
        Object {
          "author": Object {
            "profile": Object {
              "bio": "Sequi minus inventore itaque similique et.",
            },
          },
        },
        Object {
          "author": Object {
            "profile": Object {
              "bio": "Sequi minus inventore itaque similique et.",
            },
          },
        },
        Object {
          "author": Object {
            "profile": Object {
              "bio": "Sequi minus inventore itaque similique et.",
            },
          },
        },
        Object {
          "author": Object {
            "profile": Object {
              "bio": "Sequi minus inventore itaque similique et.",
            },
          },
        },
        Object {
          "author": Object {
            "profile": Object {
              "bio": "Sequi minus inventore itaque similique et.",
            },
          },
        },
      ],
    },
  },
}
`;

exports[`prisma queries with arguments and aliases 1`] = `
Object {
  "data": Object {
    "me": Object {
      "oldestPosts": Array [
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "1",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "2",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "3",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "4",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "5",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "6",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "7",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "8",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "9",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "10",
        },
      ],
      "posts": Array [
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "250",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "249",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "248",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "247",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "246",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "245",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "244",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "243",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "242",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "241",
        },
      ],
    },
  },
}
`;

exports[`prisma queries with variables and alieases 1`] = `
Object {
  "data": Object {
    "me": Object {
      "oldestPosts": Array [
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "1",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "2",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "3",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "4",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "5",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "6",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "7",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "8",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "9",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "10",
        },
      ],
      "postIds": Array [
        Object {
          "id": "250",
        },
        Object {
          "id": "249",
        },
        Object {
          "id": "248",
        },
        Object {
          "id": "247",
        },
        Object {
          "id": "246",
        },
        Object {
          "id": "245",
        },
        Object {
          "id": "244",
        },
        Object {
          "id": "243",
        },
        Object {
          "id": "242",
        },
        Object {
          "id": "241",
        },
      ],
      "posts": Array [
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "250",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "249",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "248",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "247",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "246",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "245",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "244",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "243",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "242",
        },
        Object {
          "author": Object {
            "id": "VXNlcjox",
          },
          "id": "241",
        },
      ],
    },
  },
}
`;
